contracts/lazyclaim/ERC721LazyClaim.sol

```diff
@@ -0,0 +1,159 @@
+
+import "@manifoldxyz/creator-core-solidity/contracts/core/IERC721CreatorCore.sol";
+import "@manifoldxyz/libraries-solidity/contracts/access/AdminControl.sol";
+import "@manifoldxyz/creator-core-solidity/contracts/extensions/ICreatorExtensionTokenURI.sol";
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/interfaces/IERC165.sol";
+
+import "./IERC721LazyClaim.sol";
+
+/**
+ * Lazy claim with optional whitelist ERC721 tokens
+ */
+contract ERC721LazyClaim is IERC165, IERC721LazyClaim, ICreatorExtensionTokenURI, ReentrancyGuard {
+  mapping(address => uint) public claimCounts;
+  mapping(address => mapping(uint => Claim)) public claims;
+  mapping(address => mapping(uint => uint)) public mintsPerClaim;
+  mapping(address => mapping(uint => mapping(address => uint))) public mintsPerWallet;
+  mapping(address => mapping(uint => uint)) public tokenClaims;
+
+  function supportsInterface(bytes4 interfaceId) public view virtual override(IERC165) returns (bool) {
+    return interfaceId == type(IERC721LazyClaim).interfaceId ||
+      interfaceId == type(ICreatorExtensionTokenURI).interfaceId ||
+      interfaceId == type(IERC165).interfaceId;
+  }
+
+  modifier creatorAdminRequired(address creatorContractAddress) {
+    // This is shared, not single-creator. So we must ensure that the initializer is an admin of the contract
+    AdminControl creatorCoreContract = AdminControl(creatorContractAddress);
+    require(creatorCoreContract.isAdmin(msg.sender), "Wallet is not an administrator for contract");
+    _;
+  }
+
+  // Initialize
+  function initializeClaim(
+    address creatorContractAddress,
+    bytes32 merkleRoot,
+    string calldata uri,
+    uint totalMax,
+    uint walletMax,
+    uint startDate,
+    uint endDate
+  ) external creatorAdminRequired(creatorContractAddress) {
+    // Get the index for the new claim
+    uint newIndex = claimCounts[creatorContractAddress];
+    claimCounts[creatorContractAddress] = newIndex + 1;
+
+    // Create the claim
+    claims[creatorContractAddress][newIndex] = Claim({
+      merkleRoot: merkleRoot,
+      uri: uri,
+      totalMax: totalMax,
+      walletMax: walletMax,
+      startDate: startDate,
+      endDate: endDate
+    });
+  }
+
+  // Update & Setters
+  function overwriteClaim(
+    address creatorContractAddress,
+    uint index,
+    bytes32 merkleRoot,
+    string calldata uri,
+    uint totalMax,
+    uint walletMax,
+    uint startDate,
+    uint endDate
+  ) external creatorAdminRequired(creatorContractAddress) {
+    // Overwrite the existing claim
+    claims[creatorContractAddress][index] = Claim({
+      merkleRoot: merkleRoot,
+      uri: uri,
+      totalMax: totalMax,
+      walletMax: walletMax,
+      startDate: startDate,
+      endDate: endDate
+    });
+  }
+  function setMerkleRoot(address creatorContractAddress, uint index, bytes32 merkleRoot) external {
```

Comment:
Missing permission check? Seems like anyone can set the merkleRoot.  You shuold have permission checks around each function.  My typical strategy to catch this is to look for all external and public functions, and check those.

Reply:
This is also inefficient.  Just retrieve as storage.  You don't need line 89 if you do.

Reply:
This goes for all the setX calls btw.

Reply:
Do input sanity checking on each of these.

Reply:
Removed all the setters in lieu of just the 1 overwrite method with its sanity checks